### DISCUSSION TITLE
Statically linking `torch_scatter` `torch_sparse` in C++ library

### DISCUSSION THREAD

**Discussion by ipcamit** [pyg-team/pytorch_geometric#6811] [General]
I am developing a C++ library that depends on torch scatter and sparse. Currently I am following the recommended route of linking against shared libtorchscatter.so etc. and it works fine.

However I would like to bundle torchscatter as git submodule or standalone source for better install ergonomics. So I added the libtorchscatter source in folder `torchscatter` with following line in CMakefile
```
add_subdirectory("torchscatter")
target_link_libraries(MyLib PRIVATE torchscatter PRIVATE ${libtorch})
```

I can see that it compiles without a fuss, with libtorchscatter.a being compiled and linked to my library, however when I run the test I get the dreaded error.
```
Could not find any similar ops to torch_sparse::ptr2ind. This op may not exist or may not be currently supported in TorchScript.
```

Can anyone comment or suggest a way to achieve the same? Basically I want libtorchscatter to be statically part of libMylib.so, so that it is not a runtime dependency.

**Comment by rusty1s** [comment]
Your error links to an `torch-sparse` issue. What happens if you include `torch-sparse` the same way as you include `torch-scatter`?

**Reply by ipcamit** [reply]
Sorry I think I gave a confusing picture. I did include both in compilation:
```
add_subdirectory("torchscatter")
add_subdirectory("torchspatter")
target_link_libraries(MyLib PRIVATE torchsparse PRIVATE ${libtorch})
target_link_libraries(MyLib PRIVATE torchscatter)
```
I thought I can simplify the question by providing limited details. My code compiles fine. At runtime it gives this error.

Have you tried statically linking in any of your programs? I think I might be doing something wrong in CMake files.

**Reply by rusty1s** [reply]
Mh, I am really no expert on using `libtorch`, but I assume the loaded model cannot really map to the custom ops when statically linked. I assume that your C++ code runs fine when not loading the model?

**Reply by ipcamit** [reply]
Yes. It is only when model is loaded that it creates a fuss. That does make sense. May be I can also ask this on Torch forum see what have they to say about it?

**Reply by rusty1s** [reply]
If that's possible, that would be helpful. I actually have no idea :(

**Reply by ipcamit** [reply]
I have opened the question here: https://discuss.pytorch.org/t/statically-linking-custom-ops-in-c/173839
Lets see. Thank you for your help.

**Reply by rusty1s** [reply]
What you can also try (not sure if possible), would be to build statically against `torchvision` and see if that works. If yes, there is definitely a problem within our libraries :(

**Reply by ipcamit** [reply]
OK. I have found C++ examples of torchvision models should be simple enough. Will update once I have the answer
